tests/Doctrine/Tests/ORM/Mapping/XmlMappingDriverTest.php

```diff
@@ -149,33 +149,128 @@ public function testInvalidMappingFileException()
+                $this->assertRegExp($errorMessageRegex, trim($errors[$i]->message));
+            }
+        } finally {
+            // Restore previous setting
+            libxml_clear_errors();
+            libxml_use_internal_errors($savedUseErrors);
+        }
+    }
+
+    /**
+     * @param string $xmlMappingFile
+     * @return bool
+     */
+    private function doValidateXmlSchema($xmlMappingFile)
     {
         $xsdSchemaFile  = __DIR__ . '/../../../../../doctrine-mapping.xsd';
         $dom            = new \DOMDocument('1.0', 'UTF-8');
 
         $dom->load($xmlMappingFile);
 
-        $this->assertTrue($dom->schemaValidate($xsdSchemaFile));
+        return $dom->schemaValidate($xsdSchemaFile);
     }
 
     static public function dataValidSchema()
     {
-        $list    = glob(__DIR__ . '/xml/*.xml');
-        $invalid = [
-            'Doctrine.Tests.Models.DDC889.DDC889Class.dcm'
-        ];
+        $list    = self::getAllXmlMappingPaths();
+        $invalid = self::getInvalidXmlMappingMap();
 
         $list = array_filter($list, function($item) use ($invalid){
-            return ! in_array(pathinfo($item, PATHINFO_FILENAME), $invalid);
+            return ! array_key_exists(pathinfo($item, PATHINFO_FILENAME), $invalid);
         });
 
         return array_map(function($item){
             return [$item];
         }, $list);
     }
 
+    static public function dataValidSchemaInvalidMappings()
+    {
+        $list    = self::getAllXmlMappingPaths();
+        $invalid = self::getInvalidXmlMappingMap();
+
+        $map = [];
+        foreach ($invalid as $filename => $errorMessageRegexes) {
+            $foundItem = null;
+            foreach ($list as $item) {
+                if (pathinfo($item, PATHINFO_FILENAME) === $filename) {
+                    $foundItem = $item;
+                    break;
+                }
+            }
+
+            if ($foundItem !== null) {
+                $map[$foundItem] = $errorMessageRegexes;
+            } else {
+                throw new \RuntimeException(sprintf('Found no XML mapping with filename "%s".', $filename));
+            }
+        }
+
+        return array_map(function($item, $errorMessageRegexes){
+            return [$item, $errorMessageRegexes];
+        }, array_keys($map), $map);
+    }
+
+    /**
+     * @return string[]
+     */
+    static private function getAllXmlMappingPaths()
+    {
+        return glob(__DIR__ . '/xml/*.xml');
+    }
+
+    /**
+     * @return array<string, string[]> ($filename => $errorMessageRegexes)
+     */
+    static private function getInvalidXmlMappingMap()
+    {
+        $invalid = [
+            'Doctrine.Tests.Models.DDC889.DDC889Class.dcm' => [
+                "Element '{http://doctrine-project.org/schemas/orm/doctrine-mapping}class': This element is not expected. Expected is %s.",
+            ],
+        ];
+
+        // Convert basic sprintf-style formats to PCRE patterns
```

Comment:
Maybe better: use `assertStringMatchesFormat()` instead of `assertRegeExp()` at line 180 (and e.g. `%S`, `%a` or `%A` rather than `%s` in the format)